// @aws-crypto/raw-rsa-keyring-browser/build/main/raw_rsa_keyring_web_crypto.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use
 * this file except in compliance with the License. A copy of the License is
 * located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const material_management_browser_1 = require("@aws-crypto/material-management-browser");
const web_crypto_backend_1 = require("@aws-crypto/web-crypto-backend");
const raw_keyring_1 = require("@aws-crypto/raw-keyring");
const random_source_browser_1 = require("@aws-crypto/random-source-browser");
const get_import_options_1 = require("./get_import_options");
class RawRsaKeyringWebCrypto extends material_management_browser_1.KeyringWebCrypto {
    constructor(input) {
        super();
        this._rawOnEncrypt = raw_keyring_1._onEncrypt(random_source_browser_1.randomValuesOnly);
        this._onEncrypt = async (material) => {
            const _material = await this._rawOnEncrypt(material);
            return material_management_browser_1.importForWebCryptoEncryptionMaterial(_material);
        };
        /* onDecrypt does not need to import the CryptoKey, because this is handled in the unwrap operation.
         * Encrypt needs to have access to the unencrypted data key to encrypt with other keyrings
         * but once I have functional material no other decrypt operations need to be performed.
         */
        this._onDecrypt = raw_keyring_1._onDecrypt();
        const { publicKey, privateKey, keyName, keyNamespace } = input;
        /* Precondition: RsaKeyringWebCrypto needs either a public or a private key to operate. */
        material_management_browser_1.needs(publicKey || privateKey, 'No Key provided.');
        /* Precondition: RsaKeyringWebCrypto needs identifying information for encrypt and decrypt. */
        material_management_browser_1.needs(keyName && keyNamespace, 'Identifying information must be defined.');
        const wrappingAlgorithm = get_import_options_1.getWrappingAlgorithm(publicKey, privateKey);
        const _wrapKey = async (material) => {
            /* Precondition: I must have a publicKey to wrap. */
            if (!publicKey)
                throw new Error('No publicKey configured, encrypt not supported.');
            // The nonZero backend is used because some browsers support Subtle Crypto
            // but do not support Zero Byte AES-GCM. I want to use the native
            // browser implementation of wrapKey
            const subtle = web_crypto_backend_1.getNonZeroByteBackend(await web_crypto_backend_1.getWebCryptoBackend());
            /* Can not use importCryptoKey as `wrapKey` requires extractable = true
             * In web crypto `wrapKey` is a composition of `export` and `encrypt` and
             * so the cryptoKey must have `extractable = true`.
             */
            const extractable = true;
            const { encryption } = material.suite;
            const importFormat = 'jwk';
            const keyUsages = ['wrapKey']; // limit the use of this key (*not* decrypt, encrypt, deriveKey)
            const jwk = material_management_browser_1.bytes2JWK(material_management_browser_1.unwrapDataKey(material.getUnencryptedDataKey()));
            const cryptoKey = await subtle.importKey(importFormat, jwk, encryption, extractable, keyUsages);
            const wrapFormat = 'raw';
            const encryptedArrayBuffer = await subtle.wrapKey(wrapFormat, cryptoKey, publicKey, wrappingAlgorithm);
            // Can the extractable setting of cryptoKey be changed to false here?  If so, do it.
            const edk = new material_management_browser_1.EncryptedDataKey({
                providerId: keyNamespace,
                providerInfo: keyName,
                encryptedDataKey: new Uint8Array(encryptedArrayBuffer)
            });
            return material.addEncryptedDataKey(edk, material_management_browser_1.KeyringTraceFlag.WRAPPING_KEY_ENCRYPTED_DATA_KEY);
        };
        /* returns either an array of 1 CryptoKey or an array of both from MixedBackendCryptoKey e.g.
         * [privateKey] || [nonZeroByteCryptoKey, zeroByteCryptoKey]
         */
        const privateKeys = get_import_options_1.flattenMixedCryptoKey(privateKey);
        const _unwrapKey = async (material, edk) => {
            /* Precondition: I must have a privateKey to unwrap. */
            if (!privateKey)
                throw new Error('No privateKey configured, decrypt not supported.');
            const backend = await web_crypto_backend_1.getWebCryptoBackend();
            const { suite } = material;
            const trace = {
                keyName: this.keyName,
                keyNamespace: this.keyNamespace,
                flags: material_management_browser_1.KeyringTraceFlag.WRAPPING_KEY_DECRYPTED_DATA_KEY
            };
            const format = 'raw';
            const extractable = false;
            const algorithm = suite.kdf ? suite.kdf : suite.encryption;
            const keyUsages = [material_management_browser_1.keyUsageForMaterial(material)];
            const importArgs = [
                format,
                edk.encryptedDataKey,
                privateKeys[0],
                wrappingAlgorithm,
                algorithm,
                extractable,
                keyUsages
            ];
            /* This is superior to importForWebCryptoDecryptionMaterial.
             * Here I use `subtle.unwrap` and bring the unencrypted data key into the WebCrypto world
             * without ever exposing the unencrypted data key to JavaScript.
             */
            if (web_crypto_backend_1.isFullSupportWebCryptoBackend(backend)) {
                const cryptoKey = await backend.subtle.unwrapKey(...importArgs);
                return material.setCryptoKey(cryptoKey, trace);
            }
            else {
                const importZeroBackend = [...importArgs];
                importZeroBackend[2] = privateKeys[1];
                const mixedDataKey = await Promise.all([
                    backend.nonZeroByteSubtle.unwrapKey(...importArgs),
                    backend.zeroByteSubtle.unwrapKey(...importZeroBackend)
                ]).then(([nonZeroByteCryptoKey, zeroByteCryptoKey]) => ({ nonZeroByteCryptoKey, zeroByteCryptoKey }));
                return material.setCryptoKey(mixedDataKey, trace);
            }
        };
        material_management_browser_1.readOnlyProperty(this, 'keyName', keyName);
        material_management_browser_1.readOnlyProperty(this, 'keyNamespace', keyNamespace);
        material_management_browser_1.readOnlyProperty(this, '_wrapKey', _wrapKey);
        material_management_browser_1.readOnlyProperty(this, '_unwrapKey', _unwrapKey);
    }
    _filter({ providerId, providerInfo }) {
        const { keyNamespace, keyName } = this;
        return providerId === keyNamespace && providerInfo.startsWith(keyName);
    }
    static async importPublicKey(publicKey) {
        const { wrappingAlgorithm, format, key } = get_import_options_1.getImportOptions(publicKey);
        const backend = await web_crypto_backend_1.getWebCryptoBackend();
        const subtle = web_crypto_backend_1.getNonZeroByteBackend(backend);
        return subtle.importKey(format, key, wrappingAlgorithm, false, ['wrapKey']);
    }
    static async importPrivateKey(privateKey) {
        const { wrappingAlgorithm, format, key } = get_import_options_1.getImportOptions(privateKey);
        const backend = await web_crypto_backend_1.getWebCryptoBackend();
        if (web_crypto_backend_1.isFullSupportWebCryptoBackend(backend)) {
            return backend.subtle.importKey(format, key, wrappingAlgorithm, false, ['unwrapKey']);
        }
        else {
            return Promise.all([
                backend.nonZeroByteSubtle.importKey(format, key, wrappingAlgorithm, false, ['unwrapKey']),
                backend.zeroByteSubtle.importKey(format, key, wrappingAlgorithm, false, ['unwrapKey'])
            ]).then(([nonZeroByteCryptoKey, zeroByteCryptoKey]) => ({ nonZeroByteCryptoKey, zeroByteCryptoKey }));
        }
    }
}
exports.RawRsaKeyringWebCrypto = RawRsaKeyringWebCrypto;
material_management_browser_1.immutableClass(RawRsaKeyringWebCrypto);
