// @aws-crypto/kms-keyring-browser/build/main/kms_keyring_browser.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use
 * this file except in compliance with the License. A copy of the License is
 * located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const kms_keyring_1 = require("@aws-crypto/kms-keyring");
exports.getClient = kms_keyring_1.getClient;
exports.limitRegions = kms_keyring_1.limitRegions;
exports.excludeRegions = kms_keyring_1.excludeRegions;
exports.cacheClients = kms_keyring_1.cacheClients;
const material_management_browser_1 = require("@aws-crypto/material-management-browser");
const aws_sdk_1 = require("aws-sdk"); // eslint-disable-line no-unused-vars
exports.KMS = aws_sdk_1.KMS;
const getKmsClient = kms_keyring_1.getClient(aws_sdk_1.KMS, { customUserAgent: 'AwsEncryptionSdkJavascriptBrowser' });
const cacheKmsClients = kms_keyring_1.cacheClients(getKmsClient);
exports.cacheKmsClients = cacheKmsClients;
class KmsKeyringBrowser extends kms_keyring_1.KmsKeyringClass(material_management_browser_1.KeyringWebCrypto) {
    constructor({ clientProvider = cacheKmsClients, keyIds, generatorKeyId, grantTokens, discovery } = {}) {
        super({ clientProvider, keyIds, generatorKeyId, grantTokens, discovery });
    }
    async _onEncrypt(material) {
        const _material = await super._onEncrypt(material);
        return material_management_browser_1.importForWebCryptoEncryptionMaterial(_material);
    }
    async _onDecrypt(material, encryptedDataKeys) {
        const _material = await super._onDecrypt(material, encryptedDataKeys);
        return material_management_browser_1.importForWebCryptoDecryptionMaterial(_material);
    }
}
exports.KmsKeyringBrowser = KmsKeyringBrowser;
material_management_browser_1.immutableClass(KmsKeyringBrowser);
