// @aws-crypto/kms-keyring-browser/build/module/kms_keyring_browser.js
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use
 * this file except in compliance with the License. A copy of the License is
 * located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { KmsKeyringClass, // eslint-disable-line no-unused-vars
getClient, limitRegions, excludeRegions, cacheClients } from '@aws-crypto/kms-keyring';
import { // eslint-disable-line no-unused-vars
immutableClass, importForWebCryptoEncryptionMaterial, importForWebCryptoDecryptionMaterial, KeyringWebCrypto // eslint-disable-line no-unused-vars
 } from '@aws-crypto/material-management-browser';
import { KMS } from 'aws-sdk'; // eslint-disable-line no-unused-vars
const getKmsClient = getClient(KMS, { customUserAgent: 'AwsEncryptionSdkJavascriptBrowser' });
const cacheKmsClients = cacheClients(getKmsClient);
export class KmsKeyringBrowser extends KmsKeyringClass(KeyringWebCrypto) {
    constructor({ clientProvider = cacheKmsClients, keyIds, generatorKeyId, grantTokens, discovery } = {}) {
        super({ clientProvider, keyIds, generatorKeyId, grantTokens, discovery });
    }
    async _onEncrypt(material) {
        const _material = await super._onEncrypt(material);
        return importForWebCryptoEncryptionMaterial(_material);
    }
    async _onDecrypt(material, encryptedDataKeys) {
        const _material = await super._onDecrypt(material, encryptedDataKeys);
        return importForWebCryptoDecryptionMaterial(_material);
    }
}
immutableClass(KmsKeyringBrowser);
export { getClient, cacheKmsClients, limitRegions, excludeRegions, cacheClients, KMS };
