// @aws-crypto/raw-rsa-keyring-browser/build/main/get_import_options.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use
 * this file except in compliance with the License. A copy of the License is
 * located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("./types");
const material_management_browser_1 = require("@aws-crypto/material-management-browser");
const OAEP_SHA1_MFG1 = { name: 'RSA-OAEP', hash: { name: 'SHA-1' } };
Object.freeze(OAEP_SHA1_MFG1);
Object.freeze(OAEP_SHA1_MFG1.hash);
const OAEP_SHA256_MFG1 = { name: 'RSA-OAEP', hash: { name: 'SHA-256' } };
Object.freeze(OAEP_SHA256_MFG1);
Object.freeze(OAEP_SHA256_MFG1.hash);
const OAEP_SHA384_MFG1 = { name: 'RSA-OAEP', hash: { name: 'SHA-384' } };
Object.freeze(OAEP_SHA384_MFG1);
Object.freeze(OAEP_SHA384_MFG1.hash);
const OAEP_SHA512_MFG1 = { name: 'RSA-OAEP', hash: { name: 'SHA-512' } };
Object.freeze(OAEP_SHA512_MFG1);
Object.freeze(OAEP_SHA512_MFG1.hash);
const JsonWebKeyMap = Object.freeze({
    [types_1.JsonWebKeyRsaAlg['RSA-OAEP']]: OAEP_SHA1_MFG1,
    [types_1.JsonWebKeyRsaAlg['RSA-OAEP-256']]: OAEP_SHA256_MFG1,
    [types_1.JsonWebKeyRsaAlg['RSA-OAEP-384']]: OAEP_SHA384_MFG1,
    [types_1.JsonWebKeyRsaAlg['RSA-OAEP-512']]: OAEP_SHA512_MFG1
});
const RsaPaddingMap = Object.freeze({
    [types_1.RsaPadding.OAEP_SHA1_MFG1]: OAEP_SHA1_MFG1,
    [types_1.RsaPadding.OAEP_SHA256_MFG1]: OAEP_SHA256_MFG1,
    [types_1.RsaPadding.OAEP_SHA384_MFG1]: OAEP_SHA384_MFG1,
    [types_1.RsaPadding.OAEP_SHA512_MFG1]: OAEP_SHA512_MFG1
});
function getImportOptions(keyInfo) {
    const { alg } = keyInfo;
    const { padding } = keyInfo;
    if (JsonWebKeyMap[alg]) {
        return {
            format: 'jwk',
            key: keyInfo,
            wrappingAlgorithm: JsonWebKeyMap[alg]
        };
    }
    else if (RsaPaddingMap[padding]) {
        const { format, key } = keyInfo;
        return {
            format,
            key,
            wrappingAlgorithm: RsaPaddingMap[padding]
        };
    }
    throw new Error('Unsupported RsaImportableKey');
}
exports.getImportOptions = getImportOptions;
function getWrappingAlgorithm(publicKey, privateKey) {
    const privateKeys = flattenMixedCryptoKey(privateKey);
    if (publicKey && privateKeys.length) {
        return verify(...[publicKey, ...privateKeys].map(extract));
    }
    else if (publicKey) {
        return verify(extract(publicKey));
    }
    else if (privateKeys.length) {
        return verify(...privateKeys.map(extract));
    }
    throw new Error('No Key provided.');
}
exports.getWrappingAlgorithm = getWrappingAlgorithm;
function extract(key) {
    const { algorithm } = key;
    // @ts-ignore
    const { name, hash } = algorithm;
    return { name: name, hash };
}
exports.extract = extract;
function verify(...args) {
    const [wrappingAlgorithm, ...rest] = args;
    /* Precondition: Need at least 1 algorithm to verify. */
    material_management_browser_1.needs(wrappingAlgorithm, 'Can not verify an empty set of algorithms.');
    const { name, hash } = wrappingAlgorithm;
    /* Precondition: The wrappingAlgorithm name must be a supported value. */
    material_management_browser_1.needs(types_1.JsonWebKeyRsaAlg[name], 'Algorithm name is not supported.');
    /* Precondition: The hash name must be a supported value. */
    material_management_browser_1.needs(hash && types_1.RsaHash[hash.name], 'Hash name is not supported.');
    /* Check for early return (Postcondition): Only 1 wrappingAlgorithm is clearly valid. */
    if (!rest.length)
        return wrappingAlgorithm;
    /* Precondition: All keys must have the same wrappingAlgorithm. */
    material_management_browser_1.needs(rest.every(equalWrappingAlgorithm), 'Not all RSA keys have the same wrappingAlgorithm.');
    return wrappingAlgorithm;
    function equalWrappingAlgorithm(algorithm) {
        return algorithm.name === name &&
            algorithm.hash &&
            algorithm.hash.name === hash.name;
    }
}
exports.verify = verify;
function flattenMixedCryptoKey(key) {
    /* Check for early return (Postcondition): empty inputs should return an empty array. */
    if (!key)
        return [];
    if (material_management_browser_1.isCryptoKey(key))
        return [key];
    const { nonZeroByteCryptoKey, zeroByteCryptoKey } = key;
    const keys = [nonZeroByteCryptoKey, zeroByteCryptoKey];
    /* Postcondition: Not all keys are CryptoKeys. */
    material_management_browser_1.needs(keys.every(material_management_browser_1.isCryptoKey), 'Not all keys are CryptoKeys.');
    return keys;
}
exports.flattenMixedCryptoKey = flattenMixedCryptoKey;
