// @aws-crypto/raw-rsa-keyring-browser/build/main/types.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use
 * this file except in compliance with the License. A copy of the License is
 * located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var RsaPadding;
(function (RsaPadding) {
    RsaPadding["OAEP_SHA1_MFG1"] = "OAEP_SHA1_MFG1";
    RsaPadding["OAEP_SHA256_MFG1"] = "OAEP_SHA256_MFG1";
    RsaPadding["OAEP_SHA384_MFG1"] = "OAEP_SHA384_MFG1";
    RsaPadding["OAEP_SHA512_MFG1"] = "OAEP_SHA512_MFG1";
})(RsaPadding = exports.RsaPadding || (exports.RsaPadding = {}));
var Format;
(function (Format) {
    Format["raw"] = "raw";
    Format["pkcs8"] = "pkcs8";
    Format["spki"] = "spki";
})(Format = exports.Format || (exports.Format = {}));
// RSA_PKCS1, https://github.com/aws/aws-encryption-sdk-python/blob/master/src/aws_encryption_sdk/identifiers.py#L262
var JsonWebKeyRsaAlg;
(function (JsonWebKeyRsaAlg) {
    JsonWebKeyRsaAlg["RSA-OAEP"] = "RSA-OAEP";
    JsonWebKeyRsaAlg["RSA-OAEP-256"] = "RSA-OAEP-256";
    JsonWebKeyRsaAlg["RSA-OAEP-384"] = "RSA-OAEP-384";
    JsonWebKeyRsaAlg["RSA-OAEP-512"] = "RSA-OAEP-512";
})(JsonWebKeyRsaAlg = exports.JsonWebKeyRsaAlg || (exports.JsonWebKeyRsaAlg = {}));
var RsaHash;
(function (RsaHash) {
    RsaHash["SHA-1"] = "SHA-1";
    RsaHash["SHA-256"] = "SHA-256";
    RsaHash["SHA-384"] = "SHA-384";
    RsaHash["SHA-512"] = "SHA-512";
})(RsaHash = exports.RsaHash || (exports.RsaHash = {}));
