// @aws-crypto/material-management-browser/build/main/index.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use
 * this file except in compliance with the License. A copy of the License is
 * located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./browser_cryptographic_materials_manager"));
__export(require("./material_helpers"));
__export(require("./bytes2_jwk"));
__export(require("./keyring_helpers"));
var material_management_1 = require("@aws-crypto/material-management");
exports.WebCryptoDecryptionMaterial = material_management_1.WebCryptoDecryptionMaterial;
exports.WebCryptoEncryptionMaterial = material_management_1.WebCryptoEncryptionMaterial;
exports.WebCryptoAlgorithmSuite = material_management_1.WebCryptoAlgorithmSuite;
exports.AlgorithmSuiteIdentifier = material_management_1.AlgorithmSuiteIdentifier;
exports.EncryptedDataKey = material_management_1.EncryptedDataKey;
exports.KeyringWebCrypto = material_management_1.KeyringWebCrypto;
exports.KeyringTraceFlag = material_management_1.KeyringTraceFlag;
exports.needs = material_management_1.needs;
exports.MultiKeyringWebCrypto = material_management_1.MultiKeyringWebCrypto;
exports.immutableBaseClass = material_management_1.immutableBaseClass;
exports.immutableClass = material_management_1.immutableClass;
exports.frozenClass = material_management_1.frozenClass;
exports.readOnlyProperty = material_management_1.readOnlyProperty;
exports.keyUsageForMaterial = material_management_1.keyUsageForMaterial;
exports.isValidCryptoKey = material_management_1.isValidCryptoKey;
exports.isCryptoKey = material_management_1.isCryptoKey;
exports.unwrapDataKey = material_management_1.unwrapDataKey;
