// @aws-crypto/web-crypto-backend/build/main/index.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use
 * this file except in compliance with the License. A copy of the License is
 * located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const util_locate_window_1 = require("@aws-sdk/util-locate-window");
const backend_factory_1 = require("./backend-factory");
const { getWebCryptoBackend, configureFallback } = backend_factory_1.webCryptoBackendFactory(util_locate_window_1.locateWindow());
exports.getWebCryptoBackend = getWebCryptoBackend;
exports.configureFallback = configureFallback;
var backend_factory_2 = require("./backend-factory");
exports.getNonZeroByteBackend = backend_factory_2.getNonZeroByteBackend;
exports.getZeroByteSubtle = backend_factory_2.getZeroByteSubtle;
exports.isFullSupportWebCryptoBackend = backend_factory_2.isFullSupportWebCryptoBackend;
var synchronous_random_values_1 = require("./synchronous_random_values");
exports.synchronousRandomValues = synchronous_random_values_1.synchronousRandomValues;
